File: data_structures/segment_tree.java
// A Segment Tree is a data structure used to efficiently perform range queries and updated on an array.
// It is particularly useful when you need to perform operations like finding the sum, minimum, or maximum over a range of elements,
// and you also need to update elements frequently.
// The Segment Tree allows you to do both in logarithmic time, making it a powerful tool for scenarios where both updates and queries need to be handled efficiently.

// Consider a scenario where you have an array representing daily sales in a store over a month,
// and you want to frequently query the total sales for any given week or update the sales for a specific day.
// A Segment Tree allows you to do this efficiently.

class SegmentTree {
    private int[] segmentTree;
    private int n;

    public SegmentTree(int[] arr) {
        n = arr.length;
        segmentTree = new int[2 * n];
        build(arr);
    }

    // Build the segment tree
    private void build(int[] arr) {
        // Insert leaf nodes in segment tree
        for (int i = 0; i < n; i++) {
            segmentTree[n + i] = arr[i];
        }
        // Build the tree by calculating parents
        for (int i = n; i > 0; i--) {
            segmentTree[i] = segmentTree[2 * i] + segmentTree[2 * i + 1];
        }
    }

    // Update a value in the array and segment tree
    public void update(int pos, int value) {
        pos += n;
        segmentTree[pos] = value;

        while (pos > 1) {
            pos /= 2;
            segmentTree[pos] = segmentTree[2 * pos] + segmentTree[2 * pos + 1];
        }
    }

    // Query the sum over a range (left, right)
    public int query(int left, int right) {
        left += n;
        right += n;
        int sum = 0;

        while (left < right) {
            if (left % 2 == 1) {
                sum += segmentTree[left];
                left++;
            }
            if (right % 2 == 1) {
                right--;
                sum += segmentTree[right];
            }
            left /= 2;
            right /= 2;
        }

        return sum;
    }
}

File: data_structures/segment_tree.kt
// A Segment Tree is a binary tree used for multiple purposes, most notably for:

// 1. Range query problems (like finding the minimum, maximum, or sum between two indices of an array)
// 2. Lazy propagation (deferring updates to optimize multiple updates)

// Each leaf node of the Segment Tree represents an element of the array, and each internal node represents some merging of its children nodes (e.g. sum, minimum)

// Suppose you have a series of weather stations across a long highway, each recording the temperature every hour.
// You're tasked with building a system where you can quickly find out:

// 1. The average temperature between two stations.
// 2. Update the temperature reading of a station.

// The Segment Tree can efficiently handle these operations.

class SegmentTree(private val nums: IntArray) {
    private val n = nums.size
    private val tree = IntArray(4 * n) { 0 }

    init {
        buildTree(0, n - 1, 0)
    }

    private fun buildTree(l: Int, r: Int, idx: Int) {
        if (l == r) {
            tree[idx] = nums[l]
            return
        }
        val mid = 1 + (r - 1) / 2
        buildTree(1, mid, 2 * idx + 1)
        buildTree(mid + 1, r, 2 * idx + 2)
        tree[idx] = tree[2 * idx + 1] + tree[2 * idx + 2]
    }

    fun queryRange(i: Int, j: Int): Int {
        return query(0, n -1, i, j, 0)
    }

    private fun query(l: Int, r: Int, i: Int, j: Int, idx: Int): Int {
        if (i > r || j < l) return 0
        if (i <= l && j >= r) return tree[idx]
        val mid = l + (r - 1) / 2
        return query(l, mid, i, j, 2 * idx + 1) + query(mid + 1, r, i, j, 2 * idx + 2)
    }

    fun update(i: Int, value: Int) {
        update(0, n - 1, i, value, 0)
    }

    private fun update(l: Int, r: Int, i: Int, value: Int, idx: Int) {
        if (l == r) {
            tree[idx] = value
            nums[i] = value
            return
        }
        val mid = l + (r - 1) / 2
        if (i <= mid) update(l, mid, i, value, 2 * idx + 1)
        else update(mid + 1, r, i, value, 2 * idx + 2)
        tree[idx] = tree[2 * idx + 1] + tree[2 * idx + 2]
    }
}

fun main() {
    val temperatures = intArrayOf(22, 24, 19, 23, 25, 18)
    val segmentTree = SegmentTree(temperatures)

    println(segmentTree.queryRange(1, 4)) // Sum of temperatures from index 1 to 4
    segmentTree.update(2, 20) // Update temperature at index 2
    println(segmentTree.queryRange(1, 4)) // Updated sum of temperatures from index 1 to 4
}